### PR TITLE
Merganser - Skeleton-based line detection & Bezier curves

### PR DESCRIPTION
**Pull request by tristandeleu**
## Merganser

This is a pull request for the documentation of our project **Merganser** at University of Montreal (Duckietown 2019 / IFT 6757). This project revisits the perception and state estimation pipeline with two new components:
 - A **skeleton-based line detection** pipeline to accurately detect the lines through a process called skeletonization. The pipeline was created to avoid as many false positives as possible, and can nicely handles turns.
 - A representation of the lines based on **Bezier curves**. The Bezier curves provide a compact and flexible representation to characterize any line (straight or curved). These Bezier curves, combined with a **Kalman Filter**, produce a richer estimation of the state.

This pipeline was created as ROS packages. These packages are available here:
**[https://github.com/bdura/merganser](https://github.com/bdura/merganser)**

Tristan Deleu (@tristandeleu) & Basile Dura (@bdura)

### CODE REVIEW

**Comment by tristandeleu** [comment]
Pinging @liampaull, @bhairavmehta95 & @maivincent.

Edit: If you think some of [these packages](https://github.com/bdura/merganser) might be of interest in `dt-core`, we'd be happy to discuss how we can proceed.

**Comment by bhairavmehta95** [comment]
Looks good. Unfortunately, I've not got the build system setup on my laptop, so I will leave it up to you two to make sure that the docs render correctly.

**Comment by bhairavmehta95** [comment]
@AndreaCensi can I get admin access here

**Comment by tristandeleu** [comment]
The docs renders properly on our end. The only formatting issue I am running into is how to reference figures. I tried this snippet:
```
<figure figure-id="fig:pipeline-overview">
    <figcaption>Pipeline overview</figcaption>
    <img style="width:100%" src="figures/pipeline-overview.png"/>
</figure>
```
Together with `[](#fig:pipeline-overview)`, but this seems to create an extra figure reference:
image

I was not able to find a way to do it in the Markduck documentation.

**Comment by bhairavmehta95** [comment]
Sorry, can you link the line ref?

File: book/learning_materials/50_perception_basics/31_skeletonization/31_skeletonization.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
+# Skeleton-based Line Detection {#line-detection-skeletonization status=ready}
+
+## Overview
+
+The [segment-based line detection](#line_detection) pipeline available in the [`line_detection`](https://github.com/duckietown/dt-core/tree/master19/packages/line_detector) package provides a nice and compact representation of the different lines in Duckietown for the downstream state estimation. Although it convenient, this representation does not handle turns well, and can suffer from many false positives. To improve these detections, we would like to compress the lines in a representation that preserves as much information as possible.
+
+In order to retain the morphology of the lines, we can extract their **skeletons**. In [](#fig:pipeline-overview), we show the overall pipeline to detect, filter and compress the lines. The output of this pipeline is a set of skeletons that correspond to the different lines.
+
+<figure figure-id="fig:pipeline-overview">
```
Comment (tristandeleu) [inline book/learning_materials/50_perception_basics/31_skeletonization/31_skeletonization.md]: @bhairavmehta95 Here is a reference to a figure

File: book/learning_materials/50_perception_basics/31_skeletonization/31_skeletonization.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+ - In order to keep only the detections adjacent to the road, we apply a second **flood filling** step using an initial seed in the road. This has the effect of selecting all the components connected to the road. These are shown in Figure (b). Note that this is still a binary mask.
+ - The binary mask is then used to **filter** the color detections. The filtered detections are shown in Figure (c).
+
+## Skeletonization {#skeletonization-skeletonization}
+
+The [skeletonization](https://homepages.inf.ed.ac.uk/rbf/HIPR2/skeleton.htm) is a step that removes most of the pixels in a binary image, while preserving the morphology and size of the original region. This returns the simplest pixel representation of the lines, reduced to a 1-pixel line. This representation is still in pixel space, instead of being segments, and can therefore handle curves. We use the skeletonization method described in [](#bib:lee1994skeletonization), initially introduced for 3D-surfaces, but adapted to two-dimensionial objects. The final skeletons are shown in [](#fig:skeletons).
+
+<figure figure-id="fig:skeletons">
```
Comment (tristandeleu) [inline book/learning_materials/50_perception_basics/31_skeletonization/31_skeletonization.md]: @bhairavmehta95 Here is a reference to a figure

File: book/learning_materials/50_perception_basics/31_skeletonization/31_skeletonization.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    <img style="width:180px" src="figures/skeletons.png"/>
+</figure>
+
+Note that since skeletonization retains the morphology of the original masks, it is very sensible to artifacts. Notably, if the mask contains any hole, this will be apparent in the returned skeleton, as shown in this example (bottom right).
+
+## Skeletons Clustering {#skeletonization-clustering}
+
+The skeletons detected in [](#skeletonization-skeletonization) are still live in pixel space. Even though we have segmented the lines according to their color, it might be useful to cluster them further (e.g. to distinguish between the left and right white line). For clustering, we can use the simple [DBSCAN algorithm](https://en.wikipedia.org/wiki/DBSCAN). This can be easily implemented using a combination of standard operations: a dilation with a large kernel and connected component detection. The final clusters of skeletons are shown in [](#fig:clusters).
+
+<figure figure-id="fig:clusters">
```
Comment (tristandeleu) [inline book/learning_materials/50_perception_basics/31_skeletonization/31_skeletonization.md]: @bhairavmehta95 Here is a reference to a figure

File: book/learning_materials/50_perception_basics/31_skeletonization/31_skeletonization.md (hunk 1 above)
Comment (bhairavmehta95) [reply book/learning_materials/50_perception_basics/31_skeletonization/31_skeletonization.md]: isn't this supposed to look like:

`<figure id="fig:something>`

https://raw.githubusercontent.com/duckietown/docs-learning_materials/de1c6a58f5ee3d11c088cf9579ba5958247c981a/book/learning_materials/21_systems/36_autonomy_architectures/36_autonomy_architectures.md

Comment (tristandeleu) [reply book/learning_materials/50_perception_basics/31_skeletonization/31_skeletonization.md]: Awesome, that works! Thank you!

**Comment by bhairavmehta95** [comment]
Cool, seems like it works. I don't have admin privs here, so we may just have to wait. Good job guys!

**Comment by AndreaCensi** [comment]
@bhairavmehta95 admin rights granted

**Comment by liampaull** [comment]
I (or someone) should test this and then merge into `daffy`

**Pull request merged**